File: backtracking/subsets.js
// https://leetcode.com/problems/subsets/
const dfs = (current, index, nums, result) => {
  result.push(current);
  for (let i = index; i < nums.length; i++) {
    dfs(current.concat(nums[i]), i + 1, nums, result);
  }
};
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const subsets = (nums) => {
  let result = [];
  dfs([], 0, nums, result);
  return result;
};

File: backtracking/subsets.py
# https://leetcode.com/problems/subsets/
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ret = []
        self.dfs(nums, [], ret)
        return ret

    def dfs(self, nums, path, ret):
        ret.append(path)
        for i in range(len(nums)):
            self.dfs(nums[i+1:], path+[nums[i]], ret)
